### PR TITLE
fbptcat: dump raw contents of FBPT table

### PR DESCRIPTION
**Pull request by princegeutler**
fbptcat: dump raw contents of FBPT table
FBPT =  Firmware Basic Performance Boot Table

- fbptcat is a shell command for printing the raw contents of
Firmware Basic Performance Boot Table (FBPT) within ACPI FPDT

- New FPDT package created to contain functions concerning
all tables and structures within ACPI FPDT

- New FPBT package created to contain functions concerning
the FBPT Table found within ACPI FPDT

Signed-off-by: Prince Geutler <geutlerprince@gmail.com>

### CODE REVIEW

File: cmds/contrib/fbptcat/fbptcat.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+	"github.com/u-root/u-root/pkg/acpi/fbpt"
+	"github.com/u-root/u-root/pkg/acpi/fpdt"
+)
+
+func main() {
+	// Get FPDT table from ACPI
+	var acpiFPDT acpi.Table = nil
+	var err error
+	if acpiFPDT, err = fpdt.ReadACPIFPDTTable(); err != nil {
+		fmt.Println(err)
```
Comment (rminnich) [inline cmds/contrib/fbptcat/fbptcat.go:21]: the uroot convention is to use log.Fatal.

**Review by rminnich — COMMENTED** [review]
This looks really useful, thanks!

Besides my very small nit, we're going to need tests. 

But this is a super useful tool and I hope you are ok with writing them.

**Comment by rminnich** [comment]
if would really be nice to have this, can you tidy it up and we'll try again?

**Comment by princegeutler** [comment]
> if would really be nice to have this, can you tidy it up and we'll try again?

Hi, yes thank you for the ping.

By tests do you mean tests similar to the test found in https://github.com/u-root/u-root/blob/main/pkg/acpi/linux_test.go? I am wondering if verifying table signatures would be sufficient, as testing for more would also pose a challenge for FBPT as the content can vary.

**Comment by princegeutler** [comment]
Hi, 

I have a few questions:

- I am unsure how to run the test-suite locally. Is there an easy way to achieve this? 

- I am having a bunch of commits in this pull request because I keep having to merge changes within u-root to my 'development' branch to make a new commit or amend my commits. Is this an issue? If so is there a more elegant way of modifying my pull request. 

Thank you.

**Comment by princegeutler** [comment]
> ## [Codecov](https://app.codecov.io/gh/u-root/u-root/pull/2762?src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=u-root) Report
> Attention: `111 lines` in your changes are missing coverage. Please review.
> 
> [Files](https://app.codecov.io/gh/u-root/u-root/pull/2762?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=u-root)	Coverage Δ	
> [pkg/acpi/fpdt/fpdt.go](https://app.codecov.io/gh/u-root/u-root/pull/2762?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=u-root#diff-cGtnL2FjcGkvZnBkdC9mcGR0Lmdv)	`26.66% <26.66%> (ø)`	
> [pkg/acpi/fbpt/fbpt.go](https://app.codecov.io/gh/u-root/u-root/pull/2762?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=u-root#diff-cGtnL2FjcGkvZmJwdC9mYnB0Lmdv)	`0.00% <0.00%> (ø)`	
> ... and [1 file with indirect coverage changes](https://app.codecov.io/gh/u-root/u-root/pull/2762/indirect-changes?src=pr&el=tree-more&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=u-root)
> 
> 📢 Thoughts on this report? [Let us know!](https://about.codecov.io/pull-request-comment-report/?utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=u-root).

Most of the code that I wrote in pkg/fpdt and pkg/fbpt parses the tables' contents which I can not verify since content can vary. I wrote a test to check the fpdt tables integrity by calculating the checksum (the fpbt table resides inside the fpdt table so its included in the checksum calculation). Are other tests/more code coverage needed?

**Comment by rminnich** [comment]
Well, you do need to add a DCO via git commit -s --amend for each commit

**Comment by princegeutler** [comment]
Please ignore latest changes, made a bit of a mess. Will fix later.

**Review by rminnich — APPROVED** [review]

**Review by rminnich — APPROVED** [review]

**Comment by iangoegebuer** [comment]
We have been chatting about how best to get more coverage.

At least for fbpt.go you may be able to do something similar to smbios_test.go and mock the memory ready functions. Every call to Seek and ReadFull go to a function which ultimately calls mem.Seek|ReadFull. Then you can override those functions for the test. 

I'm worried refactoring to use memio might be slower due to the number of sequential reads you need to do here and the number of times memio would have to open, seek, read, close. 

https://github.com/u-root/u-root/blob/19031e1f60ca089b4478a52abcb4de964b78f53e/pkg/smbios/smbios_test.go#L56

**Review by rminnich — APPROVED** [review]

**Pull request merged**